libraries/net/node.cpp

```diff
@@ -4919,7 +4920,7 @@ namespace graphene { namespace net { namespace detail {
         return _node_delegate->method_name(__VA_ARGS__); \
       } \
       else \
-        return _thread->async([&](){ \
+        return _thread->async([&, statistics_collector](){ \
```

Comment:
Please remove the `&` and capture `this` instead. Same below.

Reply:
I am sorry, but this won't work.

Strictly speaking, it would be enough to capture by copy only `this->_node_delegate`. But omitting the implicit by-reference capture (i.e. `&`) the `__VA_ARGS__` won't be captured.

Well, the simplest solution is to use the implicit capture by copy (i.e. `=`), which won't compile because the functions used with the macro consume parameters as references (see the interface `class node_delegate`), at least one uses such parameter as an additional output parameter.

Furthermore, it is not easily seen that the original caller provides these references as of non-stack allocated entities. Thus, to be completely bug-proof, all of these parameters shall be passed in `shared_ptr`.

Now, from the valgrind output I have produced, it does not seem this is the case. I suppose the proposed modification to be accepted as is (it provably solves an existing problem), and in the case a strictly complete solution is deemed required, a specific issue to be opened for the required refactor.

Reply:
Missed the VA_ARGS, sorry.

`handle_message`, for example, uses a stack-allocated `message` (originating from `message_oriented_connection_impl::read_loop`). Normally this lives longer than the `handle_message` call should take, however, it is re-used when reading the next message (which could lead to data corruption), and it is destroyed when reading fails (e. g. when the connection breaks down). This could even be related to #1256 @jmjatlanta ?

I agree that this should be handled in another ticket.